### PR TITLE
Add latest version to show command (option)

### PR DESCRIPTION
**Pull request by barryvdh**
This is a partial fix for #3771

Shows the recommend version, same as used in the require command. When used with the `--recommended` or `-r` flag, it shows an extra column in the table layout, or extra row for 1 package.
Multiple packages:
image

Single package:
image

Possible improvements (which I'm not really sure how to do easy):
 - Show the exact latest versions (eg. `3.1.2` instead of `^3.2`)
 - Show the current constraint? (`eg `3.x` instead of `3.1.2`)
 - Take PHP version into account?
 - Add coloring to see of the current constraint allows the latest version?

Update: Now using `--latest` / `-l` and showing the latest version, instead of best constraint.

image

### CODE REVIEW

**Comment by vinkla** [comment]
:+1:

**Comment by barryvdh** [comment]
If we want to make it more specific for versions, we could also make it a separate command, like the suggested `composer outdated`.

**Comment by vinkla** [comment]
We could also add a `--latest` or `--outdated` to the `show` command.

**Comment by barryvdh** [comment]
I've changed it to `--latest` / `-l` and used the exact version.

image

**Comment by barryvdh** [comment]
+ some coloring when the version is a match (thus latest)

image

File: src/Composer/Command/ShowCommand.php [hunk 1]
```diff
@@ -249,9 +262,11 @@ protected function execute(InputInterface $input, OutputInterface $output)
                     $input->setOption('path', false);
                 }
 
+                $latestLength = $versionLength;
```
Comment (barryvdh) [inline src/Composer/Command/ShowCommand.php]: Yes I'm making the lazy assumption that they're equal. I should probably check the actual length, but that would require some caching of the best package version, so I'll leave that as improvement for now.

**Comment by aminin** [comment]
:+1:

**Comment by barryvdh** [comment]
@Seldaek Do you think making it a separate command would be better, regarding #5060 ?

**Comment by Seldaek** [comment]
Not necessarily no. Although I guess for discoverability an outdated command that calls show --latest might be good.

**Comment by barryvdh** [comment]
Okay I'll see if I can put a deprecation notice in here also then.

**Comment by Seldaek** [comment]
Maybe best to keep it separated tbh.. Rather than having this PR grow more. At least start a new PR based on this one.

**Comment by barryvdh** [comment]
I don't think it's too intrusive on the existing code, but created a new PR, based on this one: https://github.com/composer/composer/pull/5068

**Comment by Seldaek** [comment]
Yayayay, merged :) This is very cool. I tweaked things a little (https://github.com/composer/composer/commit/e7069dd5e8ff3c2cc0fa63004f8ab0365a825037), added a `composer outdated` alias and some more coloring to follow npm style colors (orange if it needs an update but is not BC, red if it needs an update and you really should do it because it should not break). And well fixed a few things that I can't blame you for messing up because they needed pretty low level composer knowledge :) But now it handles stability flags and such properly.

**Comment by vinkla** [comment]
@seldaek great job!

Question, is it possible to exclude the packages that already are green?

**Comment by glen-84** [comment]
Maybe `list` should show all packages, and `outdated` should hide the up-to-date packages?

**Comment by Seldaek** [comment]
I now added two filters, --outdated/-o (hide greens) and --direct/-D (show only your own direct dependencies).

**Comment by rsdev000** [comment]
Hi 

Always get this error on Windows with latest php 7.
composer_error

**Comment by Seldaek** [comment]
Can you share your composer.json?

**Comment by rsdev000** [comment]
Sure. It's from a Laravel app.

```
{
    "name": "laravel/laravel",
    "description": "The Laravel Framework.",
    "keywords": ["framework", "laravel"],
    "license": "MIT",
    "type": "project",
    "require": {
        "php": ">=5.5.9",
        "laravel/framework": "5.2.*",
        "mcamara/laravel-localization": "^1.1",
        "barryvdh/laravel-httpcache": "^0.2.4",
        "radic/blade-extensions": "^6.2",
        "predis/predis": "^1.0",
        "aaronjan/housekeeper": "^2.3"
    },
    "require-dev": {
        "fzaninotto/faker": "~1.4",
        "mockery/mockery": "0.9.*",
        "phpunit/phpunit": "~4.0",
        "symfony/css-selector": "2.8.*|3.0.*",
        "symfony/dom-crawler": "2.8.*|3.0.*",
        "rap2hpoutre/laravel-log-viewer": "^0.6.0",
        "itsgoingd/clockwork": "^1.11",
        "barryvdh/laravel-ide-helper": "^2.1",
        "filp/whoops": "^2.1",
        "themsaid/laravel-langman": "^1.1",
        "sven/artisan-view": "^1.0"


    },
    "suggest": {
        "media24si/response-cache": "dev-master",
        "barryvdh/laravel-debugbar": "^2.2",
        "matthiasmullie/minify": "^1.3",
        "raveren/kint": "^1.0"
    },
    "autoload": {
        "classmap": [
            "database"
        ],
        "psr-4": {
              "App\\": "app/",
              "Larapack\\AssetCacheBuster\\": "packages/larapack/assetcachebuster/src",
              "Larapack\\DbTableGenerator\\": "packages/larapack/dbtablegenerator/src",
              "Larapack\\Logger\\": "packages/larapack/logger/src",
              "Larapack\\LoremIpsum\\": "packages/larapack/loremipsum/src",
              "Larapack\\Repository\\": "packages/larapack/repository/src"
        },
        "files": [
            "app/Lib/helpers.php"
        ]
    },
    "autoload-dev": {
        "classmap": [
            "tests/TestCase.php"
        ]
    },
    "scripts": {
        "post-root-package-install": [
            "php -r \"copy('.env.example', '.env');\""
        ],
        "post-create-project-cmd": [
            "php artisan key:generate"
        ],
        "post-install-cmd": [
            "Illuminate\\Foundation\\ComposerScripts::postInstall",
            "php artisan optimize --force"
        ],
        "post-update-cmd": [
            "Illuminate\\Foundation\\ComposerScripts::postUpdate",
            "php artisan ide-helper:generate",
            "php artisan ide-helper:meta",
            "php artisan larapack:dbtables",
            "php artisan optimize --force"
        ]
    },
    "config": {
        "preferred-install": "dist"
    }
}
```

**Comment by Seldaek** [comment]
@rsdev000 can you also show the failure if you run the command with `-vvv`? That way we get a backtrace at least and other stuff that might help figure it out.

**Comment by rsdev000** [comment]
@Seldaek yep

```
Exception trace:
 () at phar://C:/Windows/composer.phar/vendor/composer/semver/src/VersionParser.php:171
 Composer\Semver\VersionParser->normalize() at phar://C:/Windows/composer.phar/vendor/composer/semver/src/Semver.php:39
 Composer\Semver\Semver::satisfies() at C:\Users\Rui\AppData\Roaming\Composer\vendor\sllh\composer-versions-check\src\VersionsCheckPlugin.php:103
 SLLH\ComposerVersionsCheck\VersionsCheckPlugin::satisfiesComposerVersion() at C:\Users\Rui\AppData\Roaming\Composer\vendor\sllh\composer-versions-check\src\VersionsCheckPlugin.php:54
 SLLH\ComposerVersionsCheck\VersionsCheckPlugin->activate() at phar://C:/Windows/composer.phar/src/Composer/Plugin/PluginManager.php:230
 Composer\Plugin\PluginManager->addPlugin() at phar://C:/Windows/composer.phar/src/Composer/Plugin/PluginManager.php:203
 Composer\Plugin\PluginManager->registerPackage() at phar://C:/Windows/composer.phar/src/Composer/Plugin/PluginManager.php:255
 Composer\Plugin\PluginManager->loadRepository() at phar://C:/Windows/composer.phar/src/Composer/Plugin/PluginManager.php:79
 Composer\Plugin\PluginManager->loadInstalledPlugins() at phar://C:/Windows/composer.phar/src/Composer/Factory.php:357
 Composer\Factory->createComposer() at phar://C:/Windows/composer.phar/src/Composer/Factory.php:540
 Composer\Factory::create() at phar://C:/Windows/composer.phar/src/Composer/Console/Application.php:275
 Composer\Console\Application->getComposer() at phar://C:/Windows/composer.phar/src/Composer/Command/BaseCommand.php:53
 Composer\Command\BaseCommand->getComposer() at phar://C:/Windows/composer.phar/src/Composer/Command/UpdateCommand.php:97
 Composer\Command\UpdateCommand->execute() at phar://C:/Windows/composer.phar/vendor/symfony/console/Command/Command.php:259
 Symfony\Component\Console\Command\Command->run() at phar://C:/Windows/composer.phar/vendor/symfony/console/Application.php:844
 Symfony\Component\Console\Application->doRunCommand() at phar://C:/Windows/composer.phar/vendor/symfony/console/Application.php:192
 Symfony\Component\Console\Application->doRun() at phar://C:/Windows/composer.phar/src/Composer/Console/Application.php:197
 Composer\Console\Application->doRun() at phar://C:/Windows/composer.phar/vendor/symfony/console/Application.php:123
 Symfony\Component\Console\Application->run() at phar://C:/Windows/composer.phar/src/Composer/Console/Application.php:99
 Composer\Console\Application->run() at phar://C:/Windows/composer.phar/bin/composer:43
 require() at C:\Windows\composer.phar:25
```

**Comment by Seldaek** [comment]
Ah ha, so it's failing in `SLLH\ComposerVersionsCheck\VersionsCheckPlugin`, it's a bug in the plugin and not composer, see https://github.com/Soullivaneuh/composer-versions-check/issues/40

**Comment by Soullivaneuh** [comment]
@rsdev000 @Seldaek This PR should fix your issue: https://github.com/Soullivaneuh/composer-versions-check/pull/42

**Comment by Soullivaneuh** [comment]
If you are already on `v2.0.0`, here is the way to upgrade to [`v2.0.1`](https://github.com/Soullivaneuh/composer-versions-check/releases/tag/v2.0.1) and get rid of the issue:

```bash
$ composer global update --no-plugins
Changed current directory to /home/sullivan/.composer
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Removing sllh/composer-versions-check (v2.0.0)
  - Installing sllh/composer-versions-check (v2.0.1)
    Downloading: 100%         

Writing lock file
Generating autoload files
```

**Comment by Soullivaneuh** [comment]
BTW @Seldaek, it would be great to implement this kind of system (version compatibility) directly on composer. What do you think?

**Comment by rsdev000** [comment]
@Seldaek @Soullivaneuh  Thank you. Working as expected.

**Comment by glen-84** [comment]
Shouldn't `--direct` be the default for `composer outdated`? Do people usually update sub-dependencies?

**Comment by Seldaek** [comment]
@glen-84 IMO it's good to see all that's outdated because when you run update it'll typically update or if it doesn't you can then check why not and see if one of your deps needs fixing perhaps.

**Comment by glen-84** [comment]
@Seldaek Hmm. It would only update sub-deps if they fell within the version range specified in the dependency, right? However, the outdated command would show updates outside that range if I'm not mistaken.

As an example, when I run `composer outdated`, I get:

```
nikic/fast-route            v0.8.0           v1.0.0           Fast request router for PHP
```

My dependency is on `zendframework/zend-expressive-fastroute`, which specifies a version of `^0.8.0` for `nikic/fast-route`. In other words, running `composer update` would not update `nikic/fast-route`, yet it's listed as `outdated`, and there isn't much that I can do about that.

I think in most cases users are interested in their own dependencies, and not sub dependencies (which they have less control over). For this reason I think that it might make sense for this to be the default, but I'd be interested to hear what others think.

**Comment by stof** [comment]
@glen-84 depending on what the constraint is, it could or could not update it indeed. But this is true for root requirements too (except you can change those).
Knowing that a composer update could select a newer version of a dependency is still useful (to tell you to update your lock file).
And if you still get outdated packages just after an update, ``composer why-not`` is here to tell you what prevented the update (to ask the ZF team to update their requirement for instance)

**Comment by Seldaek** [comment]
I prefer to see it if it's outdated so I can check out why it's not updating and if possible go nag ZF to allow 1.0, in some cases there might be security releases and whatnot that are important to update to and might still be blocked by one of your deps' requirements. I can see that it doesn't fit everyone though, and adding a `-d` isn't the end of the world I'd think :) Changing defaults now that it's in a stable release should be avoided really, and I don't think one default can please everyone anyway.

**Comment by glen-84** [comment]
Okay, fair enough I guess.

**Pull request merged**